TESTS/mbed_hal/trng/main.cpp

```diff
@@ -0,0 +1,244 @@
+#define MSG_VALUE_LEN                   64
+#define MSG_KEY_LEN                     32
+
+#define BUFFER_LEN                      (MSG_VALUE_LEN/2)           //size of first step data, and half of the second step data
+
+#define MSG_TRNG_READY                  "ready"
+#define MSG_TRNG_BUFFER                 "buffer"
+
+#define MSG_TRNG_TEST_STEP1             "check_step1"
+#define MSG_TRNG_TEST_STEP2             "check_step2"
+#define MSG_TRNG_TEST_SUITE_ENDED       "Test_suite_ended"
+
+#define NVKEY                           1                           //NVstore key for storing and loading data
+
+/*there are some issues with nvstore and greentea reset, so for now nvstore is disabled,
+ *When  solved delete current define and replace all NVSTORE_RESET with NVSTORE_ENABLED*/
+#define NVSTORE_RESET                   (NVSTORE_ENABLED & 0)
+
+using namespace utest::v1;
+
+static int fill_buffer_trng(uint8_t *buffer, trng_t *trng_obj, size_t trng_len)
+{
+    size_t temp_size = 0, output_length = 0;
+    int trng_res = 0;
+    uint8_t *temp_in_buf = buffer;
+
+    trng_init(trng_obj);
+    memset(buffer, 0, BUFFER_LEN);
+
+    while (true) {
+        trng_res = trng_get_bytes(trng_obj, temp_in_buf, trng_len, &output_length);
+        TEST_ASSERT_EQUAL_INT_MESSAGE(0, trng_res, "trng_get_bytes error!");
+        temp_size += output_length;
+        temp_in_buf += output_length;
+        trng_len -= output_length;
+        if (temp_size >= trng_len) {
+            break;
+        }
+    }
+
+    temp_in_buf = NULL;
+    trng_free(trng_obj);
+    return 0;
+}
+
+static void compress_and_compare(char *key, char *value)
+{
+    trng_t trng_obj;
+    uint8_t out_comp_buf[BUFFER_LEN * 4] = {0}, buffer[BUFFER_LEN] = {0};
+    uint8_t input_buf[BUFFER_LEN * 2] = {0}, temp_buff[BUFFER_LEN * 2] = {0};
+    size_t comp_sz = 0;
+    unsigned int result = 0;
+
+#if NVSTORE_RESET
+    NVStore& nvstore = NVStore::get_instance();
+#endif
+
+    /*At the begining of step 2 load trng buffer from step 1*/
+    if (strcmp(key, MSG_TRNG_TEST_STEP2) == 0) {
+#if NVSTORE_RESET
+        uint16_t actual = 0;
+        result = nvstore.get(NVKEY, sizeof(buffer), buffer, actual);
+        TEST_ASSERT_EQUAL(NVSTORE_SUCCESS, result);
+#else
+        /*Using base64 to decode data sent from host*/
+        uint32_t lengthWritten = 0;
+        uint32_t charsProcessed = 0;
+        result = trng_DecodeNBase64((const char *)value, MSG_VALUE_LEN, buffer, BUFFER_LEN, &lengthWritten, &charsProcessed);
+        TEST_ASSERT_EQUAL(0, result);
+#endif
+        memcpy(input_buf, buffer, BUFFER_LEN);
+    }
+
+    /*Fill buffer with trng values*/
+    result = fill_buffer_trng(buffer, &trng_obj, sizeof(buffer));
+    TEST_ASSERT_EQUAL(0, result);
+
+    /*pithy_Compress will try to compress the random data, if it succeeded it means the data is not really random*/
+    if (strcmp(key, MSG_TRNG_TEST_STEP1) == 0) {
+        printf("\n******TRNG_TEST_STEP1*****\n");
```

Comment:
I think we can rework the amount of debug info for this test

line 140-142 could be replaced by `TEST_ASSERT_MESSAGE(condition, message);` as used on some lines in this test. Or if its worth stating the sizeof the buffer it can still be reworked to display info only if it fails.

Also remote step1,2  start/finish. If all is OK, one success should be enough rather than having 10 lines of start/finish (test case OK means it finishes and starting another one).

Reply:
@0xc0170 The amount of debug info was reduced, Because the test is comprised of several small tests I think it is important to state the steps.
If it is crucial I can omit them.

Reply:
Those small steps can be either own test cases or just run it as it is .if it passes, does not need to print anything (from test framework we get it all green). however if it fails, then these data might be crucial . Why not to print them only if something goes wrong ? This ideally would use logging that would be by default off.

@ARMmbed/mbed-os-test Please review